### PR TITLE
Add getLocksByPartitionKey()

### PR DESCRIPTION
**Pull request by lcabancla**
*Issue #, if available:* 58

*Description of changes:*

Add the following method to the AmazonDynamoDBLockClient:

```
public Stream<LockItem> getLocksByPartitionKey(String key, final boolean deleteOnRelease)
```

This method queries LockItems having the same partition key (but different sort key). It does this by calling `dynamoDB.query(this.queryRequest)`.

Currently, the only way to read multiple Lock Items from the database is by performing a scan. 
This change allows the client to query for a subset of the leases only based on the partition key.

### CODE REVIEW

File: src/main/java/com/amazonaws/services/dynamodbv2/LockItemPaginatedIterator.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.amazonaws.services.dynamodbv2;
```
Comment (kapilsingh5050) [inline src/main/java/com/amazonaws/services/dynamodbv2/LockItemPaginatedIterator.java:1]: Please add copyright notice to all newly added files

File: src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.junit.Test;
+
+import static java.util.stream.Collectors.toList;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class GetLocksByPartitionKeyTest extends InMemoryLockClientTester {
```
Comment (kapilsingh5050) [inline src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java:18]: Please add javadocs summarizing purpose of test classes added in this PR

File: src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    assertEquals(singleLock.getRecordVersionNumber(), retrievedLock.getRecordVersionNumber());
+
+    this.lockClientForRangeKeyTable.getLock(partitionKey, Optional.of(sortKey)).get().close();
+
+    locksWithParitionKey = this.lockClientForRangeKeyTable.getLocksByPartitionKey(partitionKey, deleteOnRelease).collect(toList());
+    assertEquals(Collections.emptyList(), locksWithParitionKey);
+  }
+
+  @Test
+  public void getLocksFromDynamoDB_whenMultipleLocks_shouldReturnMultipleLocks() throws LockNotGrantedException, InterruptedException, IOException {
```
Comment (kapilsingh5050) [inline src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java:72]: Let's add a similar test where there are locks with different partition key value which should not be returned

File: src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    secondLock.close();
+
+    allLocksFromDynamoDB = this.lockClientForRangeKeyTable.getLocksByPartitionKey(partitionKey, deleteOnRelease).collect(toList());
+    assertEquals(Collections.emptyList(), allLocksFromDynamoDB);
+  }
+
+  @Test
+  public void getLocksFromDynamoDB_whenMultiplePages_shouldReturnMultiplePages() throws LockNotGrantedException, InterruptedException, IOException {
+    // Scan is paginated.
```
Comment (kapilsingh5050) [inline src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java:120]: This should be changed to Query

File: src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      final LockItem acquiredLock = this.lockClientForRangeKeyTable.acquireLock(options);
+
+      acquiredLockItemsByKey.put(getLockKey(acquiredLock), acquiredLock);
+      numBytesOfData += acquiredLock.getData().get().array().length;
+    }
+
+    final boolean deleteOnRelease = false;
+    List<LockItem> locksWithPartitionKey = this.lockClientForRangeKeyTable.getLocksByPartitionKey(partitionKey, deleteOnRelease).collect(toList());
+
+    assertEquals(acquiredLockItemsByKey.size(), locksWithPartitionKey.size());
```
Comment (kapilsingh5050) [inline src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java:148]: Is there a way we can verify there were actually paginated calls to DynamoDB?

File: src/test/java/com/amazonaws/services/dynamodbv2/LockItemPaginatedQueryIteratorTest.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import org.mockito.junit.MockitoJUnitRunner;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.services.dynamodb.model.QueryRequest;
+import software.amazon.awssdk.services.dynamodb.model.QueryResponse;
+
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class LockItemPaginatedQueryIteratorTest extends TestCase {
```
Comment (kapilsingh5050) [inline src/test/java/com/amazonaws/services/dynamodbv2/LockItemPaginatedQueryIteratorTest.java:23]: Let's add more tests here. For example, next when there is a next. Also, tests for hasNext method are missing

**Comment by kapilsingh5050** [comment]
Thanks for creating this PR!

File: src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java (hunk 5 above)
Comment (lcabancla) [reply src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java:148]: Not as far as I know. Renamed the test case instead to check for complete results.

File: src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java (hunk 2 above)
Comment (lcabancla) [reply src/test/java/com/amazonaws/services/dynamodbv2/GetLocksByPartitionKeyTest.java:18]: Added.

File: src/test/java/com/amazonaws/services/dynamodbv2/LockItemPaginatedQueryIteratorTest.java (hunk 6 above)
Comment (lcabancla) [reply src/test/java/com/amazonaws/services/dynamodbv2/LockItemPaginatedQueryIteratorTest.java:43]: Added some tests.

**Pull request merged**